### PR TITLE
network: improve debug logs and add tests for IPv4LL

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

**Comment by yuwata** [comment]
@keszybz 
https://github.com/systemd/systemd/pull/17692#issuecomment-736539847

> I.e. the LL address will remain forever. Shouldn't we mark it non-preferred and drop after a minute or so?

I cannot find that behavior. The IPv4LL address is dropped soon after DHCPv4 address becomes ready. I've added a test to confirm that. PTAL.

File: src/libsystemd-network/sd-ipv4acd.c [hunk 1]
```diff
@@ -378,15 +379,33 @@ static int ipv4acd_on_packet(
...
+                return -ENODEV;
+
+        strcpy(acd->ifname, ifname);
         acd->ifindex = ifindex;
 
         return 0;
 }
 
+int sd_ipv4acd_get_ifindex(sd_ipv4acd *acd) {
+        assert_return(acd, -EINVAL);
```
Comment (keszybz) [inline src/libsystemd-network/sd-ipv4acd.c]: EBADF?

Maybe make this not assert_return but silent? See comments below...

File: src/libsystemd-network/sd-ipv4acd.c [hunk 2]
```diff
@@ -73,8 +74,10 @@ struct sd_ipv4acd {
         void* userdata;
 };
 
-#define log_ipv4acd_errno(acd, error, fmt, ...) log_internal(LOG_DEBUG, error, PROJECT_FILE, __LINE__, __func__, "IPV4ACD: " fmt, ##__VA_ARGS__)
-#define log_ipv4acd(acd, fmt, ...) log_ipv4acd_errno(acd, 0, fmt, ##__VA_ARGS__)
+#define log_ipv4acd_errno(acd, error, fmt, ...)                         \
+        log_interface_full_errno(acd ? sd_ipv4acd_get_ifname(acd) : NULL, LOG_DEBUG, error, "IPV4ACD: " fmt, ##__VA_ARGS__)
```
Comment (keszybz) [inline src/libsystemd-network/sd-ipv4acd.c]: Unfortunately this does double evaluation of `acd`. So maybe make `sd_ipv4acd_get_ifname()` just return NULL if the argument is NULL without complaining?

File: src/libsystemd-network/sd-ipv4ll.c [hunk 3]
```diff
@@ -49,8 +49,10 @@ struct sd_ipv4ll {
         void* userdata;
 };
 
-#define log_ipv4ll_errno(ll, error, fmt, ...) log_internal(LOG_DEBUG, error, PROJECT_FILE, __LINE__, __func__, "IPV4LL: " fmt, ##__VA_ARGS__)
-#define log_ipv4ll(ll, fmt, ...) log_ipv4ll_errno(ll, 0, fmt, ##__VA_ARGS__)
+#define log_ipv4ll_errno(ll, error, fmt, ...)                           \
+        log_interface_full_errno((ll && ll->acd) ? sd_ipv4acd_get_ifname(ll->acd) : NULL, LOG_DEBUG, error, "IPV4LL: " fmt, ##__VA_ARGS__)
```
Comment (keszybz) [inline src/libsystemd-network/sd-ipv4ll.c]: Same here.

File: test/test-network/systemd-networkd-tests.py [hunk 4]
```diff
@@ -4010,7 +4019,7 @@ def test_dhcp_client_with_ipv4ll_with_dhcp_server(self):
         output = check_output('ip -4 address show dev veth99 scope global dynamic')
         self.assertRegex(output, 'inet 192.168.5.[0-9]*/24 brd 192.168.5.255 scope global dynamic veth99')
         output = check_output('ip -4 address show dev veth99 scope link')
-        self.assertNotRegex(output, 'inet .* scope link')
+        self.assertNotRegex(output, 'inet 169.254.[0-9]*.[0-9]*/16 brd 169.254.255.255 scope link')
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py]: Not that it matters, but maybe:
```
r'inet 169\.254\.\d+\.\d+/16 brd 169.254.255.255 scope link'
```

File: src/network/networkd-dhcp4.c [hunk 5]
```diff
@@ -1277,6 +1277,9 @@ static int dhcp4_set_request_address(Link *link) {
         if (!a)
                 return 0;
 
+
```
Comment (keszybz) [inline src/network/networkd-dhcp4.c]: Double newline.

**Review by keszybz — COMMENTED** [review]
LGTM, except for the double-eval thing. It doesn't matter for the current usage, but I think it'd be nicer to avoid the issue to avoid bugs in the future.

I'm not sure why I saw different behaviour in my local test. I'll retest after all outstanding PRs have been merged.

File: src/libsystemd-network/sd-ipv4acd.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd-network/sd-ipv4acd.c]: EBADF?? At least under libsystemd-network, we do not use `EBADF`...

**Comment by yuwata** [comment]
@keszybz Thank you for the review. All points except `EBADF` thing are addressed. PTAL.

**Comment by keszybz** [comment]
LGTM.

**Comment by keszybz** [comment]
bionic-s390x failed in timedated test, looks unrelated.

**Comment by keszybz** [comment]
bionic-* is failing in timedated test (same as other PRs).

**Pull request merged**